Sources/SourceKitLSP/Swift/SwiftLanguageServer.swift

```diff
@@ -1461,6 +1463,72 @@ extension SwiftLanguageServer {
     // FIXME: cancellation
     _ = handle
   }
+
+  public func documentFormatting(_ req: Request<DocumentFormattingRequest>) {
+    self.queue.async {
+      do {
+        req.reply(try self.handleDocumentUpdate(req.params))
+      } catch let error as ResponseError {
+        req.reply(.failure(error))
+      } catch {
+        req.reply(.failure(ResponseError(code: .requestFailed, message: "Formatting failed: \(error.localizedDescription)")))
+      }
+    }
+  }
+
+  public func handleDocumentUpdate(_ req: DocumentFormattingRequest) throws -> [TextEdit]? {
+    guard let snapshot = documentManager.latestSnapshot(req.textDocument.uri) else {
+      return nil
+    }
+
+    var formatConfiguration: Configuration
+    if let fileURL = req.textDocument.uri.fileURL,
+        let configURL = Configuration.url(forConfigurationFileApplyingTo: fileURL),
+        let configuration = try? Configuration(contentsOf: configURL) {
+      formatConfiguration = configuration
+      // If we find a .swift-format file, we ignore the options passed to us by the editor.
+      // Most likely, the editor inferred them from the current document and thus the options
+      // passed by the editor are most likely less correct than those in .swift-format.
+    } else {
+      formatConfiguration = Configuration()
+      formatConfiguration.tabWidth = req.options.tabSize
+      formatConfiguration.indentation = req.options.insertSpaces ? .spaces(req.options.tabSize) : .tabs(1)
+    }
+
+    // Unsupported options
+    if req.options.trimTrailingWhitespace == false {
+      throw ResponseError(code: .requestFailed, message: "swift-format does not support keeping trailing whitespace; set format option to trim trailing trivia to run the formatter")
+    }
+    if req.options.insertFinalNewline == false {
+      throw ResponseError(code: .requestFailed, message: "swift-format always inserts a final newline to the file; set option to insert a final newline to run the formatter")
+    }
+    if req.options.trimFinalNewlines == false {
+      throw ResponseError(code: .requestFailed, message: "swift-format always trims final newlines; set option to trim final newlines to run the formatter")
+    }
```

Comment:
These messages are all a little confusing to me, especially the first. Is the `; set option...` really needed? If it is, can we name it explicitly with the actual value (`true`)?